### PR TITLE
Implement replace Nth command

### PR DESCRIPTION
**Pull request by rayyaw**
## Changes
- Minor Makefile improvements for QoL and to avoid recompiling the ANTLR files every time.
- Minor file handling changes.
- Implement the handling for replace Nth command.

## Testing
- Automated tests.
- Manually ran the CLI to make sure the results made sense.

## Related issues
- https://github.com/rayyaw/flash-patcher/issues/27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,8 +90,9 @@ end-patch
 remove frame_1/DoAction.as 789-1111
 ```
 
-You can use \# to write comments. There are two types of commands, `add` and `remove`. The first parameter to any command is the file to modify (in this case, "DefineSprite_1058 boss2/DoAction.as" or "frame_1/DoAction.as"). To find the name of this, export all scripts using FFDec and make a note of the file name you want to modify.
+You can use \# to write comments. There are several types of commands, `add` and `remove` being two of them. The first parameter to any command is the file to modify (in this case, "DefineSprite_1058 boss2/DoAction.as" or "frame_1/DoAction.as"). To find the name of this, export all scripts using FFDec and make a note of the file name you want to modify.
```
Comment (GTcreyon) [inline README.md]:
```suggestion
Patch files consist of a set of commands, separated onto lines. You can use \# to write comments. The first parameter to any command is the file to modify (in this case, "DefineSprite_1058 boss2/DoAction.as" or "frame_1/DoAction.as"). To find the name of this, export all scripts using FFDec and make a note of the file name you want to modify.
```
The sentence you changed here is somewhat redundant now that the commands are specified individually below. To me, it makes most sense to remove it. Doing so requires a minor rewrite for legibility, for which I have included a suggestion, but it doesn't have to be exactly as I wrote it.

File: flash_patcher/parse/visitor/asset_visitor.py [hunk 2]
```diff
@@ -22,7 +22,10 @@ def __init__(self: AssetPackProcessor, asset_folder: Path, decomp_location: Path
         self.decomp_location = decomp_location
         self.modified_scripts = set()
 
-    def visitAddAssetBlock(self, ctx: AssetPackParser.AddAssetBlockContext) -> None:
+    def visitAddAssetBlock(
```
Comment (GTcreyon) [inline flash_patcher/parse/visitor/asset_visitor.py:25]:
```suggestion
    def visit_add_asset_block(
```
Styling: Function names should be in `snake_case`. See [PEP8](https://peps.python.org/pep-0008/#function-and-variable-names) for details. That said, is this caused by antlr? I'm not sure on how that works.

File: test/inject/test_find_content.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+from flash_patcher.exception.error_manager import ErrorManager
+from flash_patcher.exception.injection import InjectionError
+from flash_patcher.inject.find_content import FindContentManager
+from flash_patcher.util.file_io import read_safe
+
+# pylint: disable=wrong-import-order
+from test.test_util.get_patch_context import get_add_patch_context, get_replace_patch_context
+
+class FindContentManagerSpec (TestCase):
+
```
Comment (GTcreyon) [inline test/inject/test_find_content.py:17]:
```suggestion
"""Some docstring."""
```
Unnecessary whitespace/missing docstring.

File: test/test_util/get_patch_context.py [hunk 4]
```diff
@@ -15,6 +15,16 @@ def get_add_patch_context(
 
     return root.addBlockHeader(offset).locationToken()
 
+def get_replace_patch_context(
+    file: Path,
+    offset: int,
+) -> PatchfileParser.LocationTokenContext:
+    root = CommonParseManager(PatchfileLexer, PatchfileParser).get_root(
+        file
+    ).replaceNthBlock(0)
```
Comment (GTcreyon) [inline test/test_util/get_patch_context.py]: Consider storing the CommonParseManager in a variable before using it, since this line is quite difficult to read and would ideally be broken up.

File: test/util/test_file_io.py [hunk 5]
```diff
@@ -1,11 +1,61 @@
...
+    @patch('pathlib.Path.open', create=True)
+    def test_rw_safe_success(
+        self: FileWritebackManagerSpec,
+        mock_open: MagicMock,
+    ) -> None:
+        mock_file = MagicMock()
+
+        mock_open.return_value.__enter__.return_value = mock_file
+
+        with open("../test/testdata/DoAction1.as", encoding="utf-8") as file:
```
Comment (GTcreyon) [inline test/util/test_file_io.py]: Consider using `Path.open()` rather than `open()`. Maybe also move the path to a constant outside of the test functions.

File: test/util/test_file_io.py [hunk 6]
```diff
@@ -1,11 +1,61 @@
...
+            content_expected = file.read()
+            mock_file.read.return_value = content_expected
```
Comment (GTcreyon) [inline test/util/test_file_io.py:28]: Question: is it safe/reasonable to dedent this line? I've always tried to follow the principle of keeping a file open for as little time as possible, but I'm not sure if that's reasonable.

Would keeping this indented perhaps have safety benefits?

**Review by GTcreyon — CHANGES_REQUESTED** [review]
Some nitpicks - all styling and formatting. Haven't tested the code yet though, so don't assume it's bug-free until re-review.

File: flash_patcher/parse/visitor/asset_visitor.py (hunk 2 above)
Comment (rayyaw) [reply flash_patcher/parse/visitor/asset_visitor.py:25]: Yes, this is an ANTLR issue. We're overriding the `visitAddAssetBlock` in the default `PatchfileVisitor` class that ANTLR generates, so there's not much we can do about this.

File: test/util/test_file_io.py (hunk 6 above)
Comment (rayyaw) [reply test/util/test_file_io.py:28]: Yes, moving this outside of the `with` block is fine. Will fix.

File: test/inject/test_find_content.py (hunk 3 above)
Comment (rayyaw) [reply test/inject/test_find_content.py:17]: This whitespace seems to be standard in our code. We should either keep it here or change it everywhere (any thoughts on which approach we should take?).

Also, all of our test case classes don't have a docstring, as the name should always follow the convention of `[base class name]Spec`, where `[base class name]` is the class whose behavior we are testing.

File: test/test_util/get_patch_context.py (hunk 4 above)
Comment (rayyaw) [reply test/test_util/get_patch_context.py]: Sounds good, I'll update our code with this!

File: README.md (hunk 1 above)
Comment (rayyaw) [reply README.md]: Sounds good, I'll rework this part of the README.

File: test/util/test_file_io.py (hunk 5 above)
Comment (rayyaw) [reply test/util/test_file_io.py]: Using `Path.open()` breaks some of the tests, as we are replacing this call with a mock. I will move the filename to a constant though.

File: test/inject/test_find_content.py (hunk 3 above)
Comment (GTcreyon) [reply test/inject/test_find_content.py:17]: Change everywhere, I'd say.

And, alright, that makes sense.

File: test/util/test_file_io.py (hunk 5 above)
Comment (GTcreyon) [reply test/util/test_file_io.py]: I don't think I know enough to understand, so I'll trust you.

**Review by GTcreyon — APPROVED** [review]
Tested with MROM, works great!
